### ISSUE TITLE
Missing accessors on PropertyDefinition

### ISSUE THREAD

**Issue by weltkante** [dotnet/corefx#4100]
The implementation of System.Reflection.Metadata [PropertyDefinition](https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/PropertyDefinition.cs#L91) and [EventDefinition](https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/EventDefinition.cs#L84) swallows some accessors and doesn't make them available publically. This has been marked as _TODO_ in the source code. These _'other'_ accessors are not often encountered, but handling them is required for certain special case tooling.

## usage of _'other'_ property accessors
* The .NET Framework exposes 'other' accessors through [PropertyInfo.GetAccessors](https://referencesource.microsoft.com/#mscorlib/system/reflection/propertyinfo.cs,496)
* .NET assemblies generated by the 'tlbimp' tool can generate this metadata when a COM property has both a PUT and a PUTREF accessor.
* They are valid metadata, even though rarely used, but if you build tools for processing or inspecting .NET assemblies (like tlbexp or ildasm do) you want to handle them.

The last point is a real-world use case I have. The tlbexp functionality built into the .NET Framework requires loading an assembly into the AppDomain to generate a TLB. Loading the assembly can fail for various reasons, so generating a TLB by only inspecting metadata is desireable. The omission of 'other' accessors is a problem here.

## Proposed API
``` csharp
namespace System.Reflection.Metadata
{
public struct PropertyAccessors 
{
    // Existing
    public MethodDefinitionHandle Getter { get; }
    public MethodDefinitionHandle Setter { get; } 

    // New
    public ImmutableArray<MethodDefinitionHandle> Others { get; }
}

public struct EventAccessors
{  
    // Existing
    public MethodDefinitionHandle Adder { get; }
    public MethodDefinitionHandle Remover { get; }
    public MethodDefinitionHandle Raiser { get; }

    // New
    public ImmutableArray<MethodDefinitionHandle> Others { get; }
}
}
```

## Details

In the large majority of cases there will be no 'other' accessors in the metadata. It is trivially possible to implement this case without any additional allocations compared to the current implementation.

For the purpose of the discussion I have provided an initial [implementation](https://github.com/weltkante/corefx/pull/3) of the suggested change.

## Updates

- Expose other accessors directly on the `PropertyAccessors` and `EventAccessors` structs instead of returning them as an out-param

**Comment by joshfree** [comment]
@weshaggard @pallavit

**Comment by karelz** [comment]
We need formal API proposal.

**Comment by weltkante** [comment]
Oh, ok, I initially submitted this as a bug, but if its not clear how to add the missing API I'll look into making a proposal and maybe even submit a possible implementation, didn't look that hard back when I noticed the bug.

**Comment by karelz** [comment]
FYI: Here's [process for new APIs](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md), incl. preferred format for proposals. Add it to this issue if/when you have it.

**Comment by karelz** [comment]
If you want to contribute to CoreFX, please check #17619 for guidance what to focus on. Thanks!

**Comment by weltkante** [comment]
@karelz I've updated the initial post and changed the bug report into a proposal request, hope I didn't miss anything.

**Comment by nguerrera** [comment]
Thanks @weltkante. I think it would be nicer without the extra overloads and out params if instead there were `Others` collections hanging off of `ProperyAccessors` and `EventAccessors`. Here's my counter proposal:

``` C#
 public struct PropertyAccessors 
{
    // Existing
    public MethodDefinitionHandle Getter { get; }
    public MethodDefinitionHandle Setter { get; } 

    // New
    public ImmutableArray<MethodDefinitionHandle> Others { get; }
}

public struct EventAccessors
{  
    // Existing
    public MethodDefinitionHandle Adder { get; }
    public MethodDefinitionHandle Remover { get;  }
    public MethodDefinitionHandle Raiser { get; } }

    // New
    public ImmutableArray<MethodDefinitionHandle> Others { get; }
}
```

Others would just be ImmutableArray<MethodDefinition>.Empty in the common case. I believe that other methods are rare enough that is fine to allocate unconditionally if there are any. That would be the simplest implementation-wise. Alternatively, the getter could do its work on demand, or it could be implemented like the other strongly typed collections that are allocation-free and just act as cursors over the backing metadata.

**Comment by karelz** [comment]
@nguerrera if you think it is ready for API review, please update top post and mark it so. If you want to double check with @weltkante first, fine with me :)

**Comment by nguerrera** [comment]
Since @weltkante is graciously driving this, I'll wait for his input on my suggestion.

**Comment by nguerrera** [comment]
Also, curious on @tmat's thoughts.

**Comment by weltkante** [comment]
Yes, having them exposed on PropertyAccessors/EventAccessor struct is nicer, I wasn't doing it because it is a struct and adding a field (i.e. making the struct larger) could have negative impact on callers and I don't know how to evaluate that for its impact.

If that isn't a concern I'm fine with moving them over.

**Comment by weltkante** [comment]
Is there anything I should do to move this forward? I'm not in a particular hurry, but I'm not sure either what the next steps would be. I assume we are waiting for feedback?

**Comment by karelz** [comment]
@nguerrera @tmat it looks like this slipped through in your area. Can you please check it out?

**Comment by tmat** [comment]
@weltkante I think we agreed above on adding  ```public ImmutableArray<MethodDefinitionHandle> Others { get; }``` to the PropertyAccessors/EventAccessors structs. The next step would be to create a PR that implements it and adds tests.

**Comment by karelz** [comment]
Yep @nguerrera was just waiting on your opinion: https://github.com/dotnet/corefx/issues/4100#issuecomment-292672561

If you (the owners) are ok with the API shape, please update the top post with final API shape and mark it for API review (api-ready-for-review). Once it gets approved, it can be implemented.

**Comment by tmat** [comment]
@karelz My opinion was expressed by 👍 on Nick's proposal.

**Comment by karelz** [comment]
Well, not sure when you added it, but as you see it was not clear ;-)

Also you still didn't mark the bug for next stage (API review) - which should have happen once area experts/owners make the decision ...

**Comment by tmat** [comment]
Updated the proposal and marked for API review.

**Comment by weltkante** [comment]
Thanks, I've implemented the updated API (link in top level post is updated) and also have added a test.

**Comment by karelz** [comment]
We're blocked on API review -- see [API review process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md). It will happen in Tue meeting, either next week or some week after (depends on how many API proposals we will have to review).

**Comment by terrajobst** [comment]
Looks good. Just one question, it seems the IL spec also allows `.custom`. How is that represented?

**Comment by karelz** [comment]
Pending the question, are you @weltkante interested in grabbing the issue?

**Comment by weltkante** [comment]
Taking @terrajobst edit into account, I'd assume they are already exposed via GetCustomAttributes, so there is probably nothing more to do in this regard.

@karelz if you mean this issue, yes. I've already addressed the comments on my local forks PR, so I assume (according to your [linked document](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md)) I now make a new PR against corefx. I'll look into that tomorrow.

**Comment by karelz** [comment]
Great, thanks! Assigned to you. Also sent you collaborator invite - pro tip: Disable monitoring entire repo, it's a lot ;).
